File: node/deploy/src/wallet.js
const { ec: EC } = require('elliptic');
const crypto = require('crypto');
const ec = new EC('secp256k1');
const fs = require('fs');
const path = require('path');

// Load BIP39 wordlist
const wordlist = require('./wordlist.json');

// Security constants
const PBKDF2_ITERATIONS = 600000; // Increased from 100000
const PBKDF2_DIGEST = 'sha512'; // Upgraded from sha256
const ENCRYPTION_ALGORITHM = 'aes-256-gcm';
const MIN_PASSWORD_LENGTH = 12;

class Wallet {
    constructor(privateKey = null) {
        this.keyPair = privateKey ? ec.keyFromPrivate(privateKey) : ec.genKeyPair();
        this.publicKey = this.keyPair.getPublic('hex');
        this.privateKey = this.keyPair.getPrivate('hex');
    }

    getAddress() {
        return crypto.createHash('ripemd160')
            .update(Buffer.from(this.publicKey, 'hex'))
            .digest('hex');
    }

    getBalance(blockchain) {
        if (!blockchain || !blockchain.chain) {
            throw new Error('Invalid blockchain object');
        }
        
        let balance = 0;
        for (const block of blockchain.chain) {
            if (block.data && block.data.validator === this.publicKey) {
                balance += block.data.reward || 0;
            }
        }
        return balance;
    }

    sign(data) {
        if (!data) {
            throw new Error('Cannot sign empty data');
        }
        return this.keyPair.sign(data).toDER('hex');
    }

    verify(data, signature) {
        if (!data || !signature) {
            throw new Error('Data and signature are required for verification');
        }
        return this.keyPair.verify(data, signature);
    }

    // Generate a backup phrase (24 words) using secure entropy
    generateBackupPhrase() {
        // Use 32 bytes (256 bits) of secure entropy
        const entropy = crypto.randomBytes(32);
        const words = [];
        
        // Use a cryptographically secure method to derive word indices
        const hash = crypto.createHash('sha256').update(entropy).digest();
        
        for (let i = 0; i < 24; i++) {
            // Use 11 bits per word as per BIP39 standard
            const position = i * 11;
            const bytePos = Math.floor(position / 8);
            const bitPos = position % 8;
            
            // Extract 11 bits for each word
            let value = 0;
            if (bitPos <= 5) {
                value = (hash[bytePos] & (0xff >> bitPos)) << (11 - (8 - bitPos));
                value |= (hash[bytePos + 1] & 0xff) >> (8 - (11 - (8 - bitPos)));
            } else {
                value = (hash[bytePos] & (0xff >> bitPos)) << (8 + (11 - (8 - bitPos)));
                value |= ((hash[bytePos + 1] & 0xff) << (11 - 8)) >> (bitPos - 5);
            }
            
            // Ensure index is within bounds
            const index = value % wordlist.length;
            words.push(wordlist[index]);
        }

        return words.join(' ');
    }

    // Encrypt wallet data with a password
    encrypt(password) {
        if (!password || typeof password !== 'string') {
            throw new Error('Password must be a non-empty string');
        }
        
        if (password.length < MIN_PASSWORD_LENGTH) {
            throw new Error(`Password must be at least ${MIN_PASSWORD_LENGTH} characters long`);
        }

        const salt = crypto.randomBytes(32); // Increased from 16
        const key = crypto.pbkdf2Sync(password, salt, PBKDF2_ITERATIONS, 32, PBKDF2_DIGEST);
        const iv = crypto.randomBytes(16);

        const cipher = crypto.createCipheriv(ENCRYPTION_ALGORITHM, key, iv);
        
        const walletData = JSON.stringify({
            privateKey: this.privateKey,
            publicKey: this.publicKey,
            address: this.getAddress(),
            createdAt: new Date().toISOString()
        });

        let encrypted = cipher.update(walletData, 'utf8', 'hex');
        encrypted += cipher.final('hex');
        const authTag = cipher.getAuthTag();

        return {
            encrypted,
            salt: salt.toString('hex'),
            iv: iv.toString('hex'),
            authTag: authTag.toString('hex'),
            algorithm: ENCRYPTION_ALGORITHM,
            kdf: {
                function: 'pbkdf2',
                iterations: PBKDF2_ITERATIONS,
                digest: PBKDF2_DIGEST
            }
        };
    }

    // Save encrypted wallet to file with path sanitization
    save(password, filename = 'wallet.json') {
        if (!password) {
            throw new Error('Password is required');
        }
        
        // Sanitize filename to prevent path traversal
        const sanitizedFilename = path.basename(filename.replace(/[^a-zA-Z0-9._-]/g, '_'));
        
        if (!sanitizedFilename.endsWith('.json')) {
            throw new Error('Wallet filename must have .json extension');
        }
        
        const encryptedData = this.encrypt(password);
        
        const walletDir = path.join(process.cwd(), 'wallets');
        if (!fs.existsSync(walletDir)) {
            fs.mkdirSync(walletDir, { mode: 0o700 }); // Secure permissions
        }

        const walletPath = path.join(walletDir, sanitizedFilename);
        
        // Use atomic write to prevent partial writes
        const tempPath = `${walletPath}.tmp`;
        fs.writeFileSync(
            tempPath,
            JSON.stringify(encryptedData, null, 2),
            { mode: 0o600 } // Secure file permissions
        );
        fs.renameSync(tempPath, walletPath);

        return {
            address: this.getAddress(),
            publicKey: this.publicKey,
            path: walletPath
        };
    }

    // Decrypt and load wallet from file with security checks
    static load(password, filename = 'wallet.json') {
        try {
            if (!password) {
                throw new Error('Password is required');
            }
            
            // Sanitize filename to prevent path traversal
            const sanitizedFilename = path.basename(filename.replace(/[^a-zA-Z0-9._-]/g, '_'));
            
            const walletPath = path.join(process.cwd(), 'wallets', sanitizedFilename);
            
            // Check if file exists
            if (!fs.existsSync(walletPath)) {
                throw new Error(`Wallet file not found: ${sanitizedFilename}`);
            }
            
            // Read with proper error handling
            let fileContent;
            try {
                fileContent = fs.readFileSync(walletPath, 'utf8');
            } catch (err) {
                throw new Error(`Error reading wallet file: ${err.message}`);
            }
            
            let encryptedData;
            try {
                encryptedData = JSON.parse(fileContent);
            } catch (err) {
                throw new Error(`Invalid wallet file format: ${err.message}`);
            }

            // Validate required fields
            if (!encryptedData.salt || !encryptedData.iv || !encryptedData.authTag || !encryptedData.encrypted) {
                throw new Error('Wallet file is missing required encryption data');
            }

            const salt = Buffer.from(encryptedData.salt, 'hex');
            const iv = Buffer.from(encryptedData.iv, 'hex');
            const authTag = Buffer.from(encryptedData.authTag, 'hex');
            
            // Use stored KDF parameters if available, otherwise use defaults
            const iterations = encryptedData.kdf?.iterations || PBKDF2_ITERATIONS;
            const digest = encryptedData.kdf?.digest || PBKDF2_DIGEST;
            
            const key = crypto.pbkdf2Sync(password, salt, iterations, 32, digest);
            
            const algorithm = encryptedData.algorithm || ENCRYPTION_ALGORITHM;
            const decipher = crypto.createDecipheriv(algorithm, key, iv);
            decipher.setAuthTag(authTag);

            let decrypted;
            try {
                decrypted = decipher.update(encryptedData.encrypted, 'hex', 'utf8');
                decrypted += decipher.final('utf8');
            } catch (err) {
                throw new Error('Invalid password or corrupted wallet file');
            }

            let walletData;
            try {
                walletData = JSON.parse(decrypted);
            } catch (err) {
                throw new Error('Corrupted wallet data');
            }
            
            // Validate wallet data
            if (!walletData.privateKey || !walletData.publicKey) {
                throw new Error('Wallet data is missing required fields');
            }

            return new Wallet(walletData.privateKey);
        } catch (error) {
            console.error('Error loading wallet:', error.message);
            return null;
        }
    }
}

module.exports = { Wallet };

File: blockchain/wallet.py
from Crypto.PublicKey import RSA
from Crypto.Cipher import PKCS1_OAEP, AES
from Crypto.Hash import SHA256
from Crypto.Random import get_random_bytes
from Crypto.Util.Padding import pad, unpad
from Crypto.Signature import pkcs1_15
import base64
import json
import os
from mnemonic import Mnemonic
import secrets
import time
from typing import Tuple, Optional, Dict, Any
import structlog
import json
import shutil

logger = structlog.get_logger()

# Constants
SATOSHI = 0.00000001  # Smallest unit of BT2C (1 satoshi)
WALLET_DIR = os.path.expanduser("~/.bt2c/wallets")

# Security constants
MIN_PASSWORD_LENGTH = 12
MIN_ENTROPY_BITS = 128  # Minimum password entropy (combination of length and character types)
PBKDF2_ITERATIONS = 600000  # Number of iterations for PBKDF2
PBKDF2_DIGEST = 'sha512'  # Hash algorithm for PBKDF2

class Wallet:
    def __init__(self):
        self.private_key = None
        self.public_key = None
        self.address = None
        self.seed_phrase = None
        self.key_created_at = int(time.time())
        self.key_rotation_interval = 90 * 24 * 60 * 60  # 90 days in seconds
    
    @classmethod
    def generate(cls, seed_phrase=None):
        """
        Generate a new wallet with RSA key pair
        """
        try:
            wallet = cls()
            
            # Generate mnemonic if not provided
            if not seed_phrase:
                # Ensure sufficient entropy (256 bits) for BIP39 seed phrase
                entropy = secrets.token_bytes(32)  # 256 bits of entropy
                m = Mnemonic("english")
                seed_phrase = m.to_mnemonic(entropy)
            
            wallet.seed_phrase = seed_phrase
            
            # Create RSA key pair
            private_key = RSA.generate(2048)
            public_key = private_key.publickey()
            
            wallet.private_key = private_key
            wallet.public_key = public_key
            
            # Generate address from public key
            wallet.address = wallet._generate_address(public_key)
            
            return wallet
        except Exception as e:
            logger.error("wallet_generation_failed", error=str(e))
            raise ValueError(f"Failed to generate wallet: {str(e)}")
    
    def _generate_address(self, public_key):
        """
        Generate a BT2C address from a public key
        """
        public_key_bytes = public_key.export_key('DER')
        address_hash = SHA256.new(public_key_bytes).digest()
        # Remove padding characters (=) from base32 encoding
        b32_encoded = base64.b32encode(address_hash[:16]).decode('utf-8').lower().rstrip('=')
        return "bt2c_" + b32_encoded
    
    def sign(self, data):
        """
        Sign data with the wallet's private key
        """
        if not self.private_key:
            raise ValueError("No private key available for signing")
        
        # Validate input
        if not isinstance(data, (bytes, str)):
            raise TypeError("Data must be bytes or string")
            
        if isinstance(data, str):
            data = data.encode('utf-8')
            
        h = SHA256.new(data)
        signature = pkcs1_15.new(self.private_key).sign(h)
        return base64.b64encode(signature).decode('utf-8')
    
    def save(self, filename, password):
        """
        Save wallet to file with password encryption
        """
        if not self.private_key:
            raise ValueError("No wallet data to save")
            
        # Validate password strength
        if not password or len(password) < MIN_PASSWORD_LENGTH:
            raise ValueError(f"Password must be at least {MIN_PASSWORD_LENGTH} characters")
            
        # Check password entropy
        import re
        has_upper = bool(re.search(r'[A-Z]', password))
        has_lower = bool(re.search(r'[a-z]', password))
        has_digit = bool(re.search(r'\d', password))
        has_special = bool(re.search(r'[^A-Za-z0-9]', password))
        
        # Calculate character set size based on character types used
        char_types_count = sum([has_upper, has_lower, has_digit, has_special])
        
        # Require at least 3 character types for strong passwords
        if char_types_count < 3:
            raise ValueError(
                f"Password too weak. Please use a stronger password with a mix of at least 3 types: "
                f"uppercase, lowercase, digits, and special characters."
            )
            
        # For passwords meeting character type requirements, ensure sufficient length
        if len(password) < MIN_PASSWORD_LENGTH:
            raise ValueError(f"Password must be at least {MIN_PASSWORD_LENGTH} characters")
            
        # Prevent path traversal
        safe_filename = os.path.basename(filename)
        if safe_filename != filename:
            raise ValueError("Invalid filename: path traversal detected")
            
        # Create wallet directory if it doesn't exist
        os.makedirs(WALLET_DIR, exist_ok=True)
        
        # Export private key
        private_key_data = self.private_key.export_key('PEM')
        
        # Generate salt and encryption key using PBKDF2
        salt = get_random_bytes(32)  # Increased from 16 to 32 bytes
        
        # Use PBKDF2 for key derivation with high iteration count
        from Crypto.Protocol.KDF import PBKDF2
        from Crypto.Hash import SHA512
        
        # Use the appropriate hash module based on PBKDF2_DIGEST
        if PBKDF2_DIGEST == 'sha512':
            hash_module = SHA512
        else:
            # Default to SHA512 if specified digest is not available
            hash_module = SHA512
            
        key = PBKDF2(
            password=password.encode('utf-8'),
            salt=salt,
            dkLen=32,  # 256-bit key
            count=PBKDF2_ITERATIONS,
            hmac_hash_module=hash_module
        )
        
        # Create AES cipher
        cipher = AES.new(key, AES.MODE_CBC)
        iv = cipher.iv
        
        # Encrypt the private key
        encrypted_data = cipher.encrypt(pad(private_key_data, AES.block_size))
        
        # Prepare wallet data
        wallet_data = {
            'address': self.address,
            'encrypted_key': base64.b64encode(encrypted_data).decode('utf-8'),
            'salt': base64.b64encode(salt).decode('utf-8'),
            'iv': base64.b64encode(iv).decode('utf-8'),
            'seed_phrase_hint': self.seed_phrase.split()[0] if self.seed_phrase else None
        }
        
        # Write to file atomically
        filepath = os.path.join(WALLET_DIR, safe_filename)
        temp_filepath = filepath + '.tmp'
        
        with open(temp_filepath, 'w') as f:
            json.dump(wallet_data, f)
            f.flush()
            os.fsync(f.fileno())
            
        os.rename(temp_filepath, filepath)
        
        # Set secure file permissions
        os.chmod(filepath, 0o600)
        
        return filepath
    
    @classmethod
    def load(cls, filename, password):
        """
        Load wallet from file with password decryption
        """
        # Prevent path traversal
        safe_filename = os.path.basename(filename)
        if safe_filename != filename:
            raise ValueError("Invalid filename: path traversal detected")
            
        filepath = os.path.join(WALLET_DIR, safe_filename)
        
        try:
            with open(filepath, 'r') as f:
                wallet_data = json.load(f)
                
            # Extract data
            address = wallet_data['address']
            encrypted_data = base64.b64decode(wallet_data['encrypted_key'])
            salt = base64.b64decode(wallet_data['salt'])
            iv = base64.b64decode(wallet_data['iv'])
            
            # Derive key from password using PBKDF2
            from Crypto.Protocol.KDF import PBKDF2
            from Crypto.Hash import SHA512
            
            # Use the appropriate hash module based on PBKDF2_DIGEST
            if PBKDF2_DIGEST == 'sha512':
                hash_module = SHA512
            else:
                # Default to SHA512 if specified digest is not available
                hash_module = SHA512
                
            key = PBKDF2(
                password=password.encode('utf-8'),
                salt=salt,
                dkLen=32,  # 256-bit key
                count=PBKDF2_ITERATIONS,
                hmac_hash_module=hash_module
            )
            
            # Create AES cipher
            cipher = AES.new(key, AES.MODE_CBC, iv)
            
            # Decrypt private key
            private_key_data = unpad(cipher.decrypt(encrypted_data), AES.block_size)
            
            # Create wallet
            wallet = cls()
            wallet.private_key = RSA.import_key(private_key_data)
            wallet.public_key = wallet.private_key.publickey()
            wallet.address = address
            
            return wallet
        except FileNotFoundError:
            raise ValueError(f"Wallet file not found: {filename}")
        except (ValueError, KeyError) as e:
            raise ValueError(f"Invalid wallet file or password: {str(e)}")
        except Exception as e:
            logger.error("wallet_load_failed", error=str(e), filename=filename)
            raise ValueError(f"Failed to load wallet: {str(e)}")
    
    @classmethod
    def recover(cls, seed_phrase, password):
        """
        Recover a wallet from a seed phrase
        """
        try:
            wallet = cls.generate(seed_phrase)
            
            # Save the wallet with the provided password
            filename = wallet.address + ".json"
            wallet.save(filename, password)
            
            logger.info("wallet_recovered", address=wallet.address)
            return wallet
        except Exception as e:
            logger.error("wallet_recovery_failed", error=str(e))
            raise ValueError(f"Failed to recover wallet: {str(e)}")
    
    @classmethod
    def create(cls, password):
        """
        Create a new wallet and save it
        """
        try:
            wallet = cls.generate()
            
            # Save the wallet with the provided password
            filename = wallet.address + ".json"
            wallet.save(filename, password)
            
            logger.info("wallet_created", address=wallet.address)
            return wallet, wallet.seed_phrase
        except Exception as e:
            logger.error("wallet_creation_failed", error=str(e))
            raise ValueError(f"Failed to create wallet: {str(e)}")
            
    def rotate_keys(self, password, new_password=None):
        """
        Rotate wallet keys for enhanced security.
        
        This creates a new key pair while preserving the wallet address and seed phrase.
        The old key is securely wiped from memory after rotation.
        
        Args:
            password: Current wallet password for verification
            new_password: Optional new password to use (if changing password during rotation)
            
        Returns:
            True if rotation was successful, raises exception otherwise
        """
        try:
            if not self.private_key or not self.seed_phrase:
                raise ValueError("Cannot rotate keys: wallet not fully initialized")
                
            # Verify current password by attempting to export and re-import the private key
            try:
                # Export private key with current password as a test
                private_key_data = self.private_key.export_key('PEM')
            except Exception:
                raise ValueError("Invalid current password")
                
            # Generate new key pair using the same seed phrase
            new_wallet = self.generate(self.seed_phrase)
            
            # Store old address for verification
            old_address = self.address
            
            # Backup old keys securely
            old_private_key = self.private_key
            old_public_key = self.public_key
            
            # Update keys
            self.private_key = new_wallet.private_key
            self.public_key = new_wallet.public_key
            
            # Verify address hasn't changed (should be deterministic from seed phrase)
            if self.address != old_address:
                # Restore old keys if address changed unexpectedly
                self.private_key = old_private_key
                self.public_key = old_public_key
                raise ValueError("Key rotation failed: address mismatch")
                
            # Update key creation timestamp
            self.key_created_at = int(time.time())
            
            # Save with new password if provided, otherwise use current password
            save_password = new_password if new_password else password
            filename = self.address + ".json"
            self.save(filename, save_password)
            
            # Securely wipe old keys from memory
            # This is a simplified version - in production, use secure memory wiping
            if old_private_key:
                del old_private_key
            if old_public_key:
                del old_public_key
                
            logger.info("wallet_keys_rotated", address=self.address)
            return True
            
        except Exception as e:
            logger.error("key_rotation_failed", error=str(e))
            raise ValueError(f"Failed to rotate keys: {str(e)}")
            
    def needs_rotation(self):
        """
        Check if keys need rotation based on age.
        
        Returns:
            True if keys are older than the rotation interval, False otherwise
        """
        current_time = int(time.time())
        key_age = current_time - self.key_created_at
        return key_age > self.key_rotation_interval
        
    def create_secure_backup(self, backup_path: str, password: str) -> str:
        """
        Create a secure encrypted backup of the wallet.
        
        This creates a full backup including:
        - Private key (encrypted)
        - Public key
        - Address
        - Seed phrase (encrypted)
        - Key creation timestamp
        
        Args:
            backup_path: Directory to store the backup
            password: Password to encrypt the backup
            
        Returns:
            Path to the backup file
        """
        if not self.private_key or not self.seed_phrase:
            raise ValueError("Cannot backup: wallet not fully initialized")
            
        try:
            # Create backup directory if it doesn't exist
            os.makedirs(backup_path, exist_ok=True)
            
            # Generate a timestamp for the backup filename
            timestamp = time.strftime("%Y%m%d-%H%M%S")
            backup_filename = f"{self.address}-{timestamp}.backup"
            backup_file_path = os.path.join(backup_path, backup_filename)
            
            # Prepare backup data
            backup_data = {
                "address": self.address,
                "key_created_at": self.key_created_at,
                "backup_created_at": int(time.time()),
                "version": "1.0"
            }
            
            # Generate salt and encryption key for the backup
            salt = get_random_bytes(32)
            
            # Use PBKDF2 for key derivation
            from Crypto.Protocol.KDF import PBKDF2
            from Crypto.Hash import SHA512
            
            # Use the appropriate hash module based on PBKDF2_DIGEST
            if PBKDF2_DIGEST == 'sha512':
                hash_module = SHA512
            else:
                # Default to SHA512 if specified digest is not available
                hash_module = SHA512
                
            key = PBKDF2(
                password=password.encode('utf-8'),
                salt=salt,
                dkLen=32,
                count=PBKDF2_ITERATIONS,
                hmac_hash_module=hash_module
            )
            
            # Create AES cipher
            iv = get_random_bytes(16)
            cipher = AES.new(key, AES.MODE_CBC, iv)
            
            # Encrypt private key and seed phrase
            private_key_data = self.private_key.export_key('PEM')
            seed_data = self.seed_phrase.encode('utf-8')
            
            # Combine private key and seed phrase with a separator
            combined_data = private_key_data + b"||SEPARATOR||" + seed_data
            
            # Pad and encrypt
            padded_data = pad(combined_data, AES.block_size)
            encrypted_data = cipher.encrypt(padded_data)
            
            # Add encrypted data to backup
            backup_data["encrypted_data"] = base64.b64encode(encrypted_data).decode('utf-8')
            backup_data["salt"] = base64.b64encode(salt).decode('utf-8')
            backup_data["iv"] = base64.b64encode(iv).decode('utf-8')
            
            # Save backup file
            with open(backup_file_path, 'w') as f:
                json.dump(backup_data, f, indent=2)
                
            logger.info("wallet_backup_created", address=self.address, backup_file=backup_file_path)
            return backup_file_path
            
        except Exception as e:
            logger.error("wallet_backup_failed", error=str(e))
            raise ValueError(f"Failed to create backup: {str(e)}")
            
    @classmethod
    def restore_from_backup(cls, backup_file_path: str, password: str):
        """
        Restore a wallet from a secure backup file.
        
        Args:
            backup_file_path: Path to the backup file
            password: Password to decrypt the backup
            
        Returns:
            Restored Wallet instance
        """
        try:
            # Read backup file
            with open(backup_file_path, 'r') as f:
                backup_data = json.load(f)
                
            # Extract data
            address = backup_data['address']
            key_created_at = backup_data.get('key_created_at', int(time.time()))
            encrypted_data = base64.b64decode(backup_data['encrypted_data'])
            salt = base64.b64decode(backup_data['salt'])
            iv = base64.b64decode(backup_data['iv'])
            
            # Derive key from password
            from Crypto.Protocol.KDF import PBKDF2
            from Crypto.Hash import SHA512
            
            # Use the appropriate hash module based on PBKDF2_DIGEST
            if PBKDF2_DIGEST == 'sha512':
                hash_module = SHA512
            else:
                # Default to SHA512 if specified digest is not available
                hash_module = SHA512
                
            key = PBKDF2(
                password=password.encode('utf-8'),
                salt=salt,
                dkLen=32,
                count=PBKDF2_ITERATIONS,
                hmac_hash_module=hash_module
            )
            
            # Create AES cipher
            cipher = AES.new(key, AES.MODE_CBC, iv)
            
            # Decrypt data
            decrypted_data = unpad(cipher.decrypt(encrypted_data), AES.block_size)
            
            # Split private key and seed phrase
            parts = decrypted_data.split(b"||SEPARATOR||")
            if len(parts) != 2:
                raise ValueError("Invalid backup format")
                
            private_key_data = parts[0]
            seed_phrase = parts[1].decode('utf-8')
            
            # Create wallet
            wallet = cls()
            wallet.private_key = RSA.import_key(private_key_data)
            wallet.public_key = wallet.private_key.publickey()
            wallet.seed_phrase = seed_phrase
            wallet.address = address
            wallet.key_created_at = key_created_at
            
            # Verify address matches
            generated_address = wallet._generate_address(wallet.public_key)
            if generated_address != address:
                raise ValueError("Address mismatch: backup may be corrupted")
                
            logger.info("wallet_restored_from_backup", address=wallet.address)
            return wallet
            
        except FileNotFoundError:
            raise ValueError(f"Backup file not found: {backup_file_path}")
        except json.JSONDecodeError:
            raise ValueError(f"Invalid backup file format: {backup_file_path}")
        except Exception as e:
            logger.error("wallet_restore_failed", error=str(e))
            raise ValueError(f"Failed to restore wallet: {str(e)}")

    def get_public_key_from_address(self, address: str):
        """
        Get the public key from an address.
        This is a simplified implementation for testing purposes.
        In a real implementation, this would require a lookup in a database or blockchain.
        
        Args:
            address: The wallet address
            
        Returns:
            The public key if found, None otherwise
        """
        # For testing purposes, we'll just return our own public key
        # In a real implementation, this would require a lookup
        if self.address == address and self.public_key:
            return self.public_key
            
        # For the genesis address, return a dummy public key
        if address == "0" * 64:
            # Create a dummy RSA key for the genesis address
            from Crypto.PublicKey import RSA
            return RSA.generate(2048).publickey()
            
        # For funded wallets in tests, we can use this as a fallback
        # This is not secure and should not be used in production
        try:
            # Try to extract a seed from the address (this is just for testing)
            import hashlib
            seed = hashlib.sha256(address.encode()).digest()
            from Crypto.PublicKey import RSA
            key = RSA.generate(2048, lambda n: int.from_bytes(seed[:4], 'big'))
            return key.publickey()
        except:
            return None
